unit/source_unit.go

```diff
@@ -77,12 +77,33 @@ type SourceUnit struct {
 	// automatically according to the user's configuration.
 	Ops map[string]*srclib.ToolRef
 
+	// CachedRev is the vcs revision that contains the last modified version
+	// of this source unit.
+	CachedRev string
+
 	// TODO(sqs): add a way to specify the toolchains and tools to use for
 	// various tasks on this source unit
 }
 
 //END SourceUnit OMIT
 
+// Doesn't do globs.
+func (u SourceUnit) ContainsAny(filenames []string) bool {
```

Comment:
This changeset doesn't handle deleted files. If a file is deleted from
a source unit, the source unit will only be recomputed if another file
that belongs to the source unit is changed. It makes sense to fix this
eventually, but I didn't think it was a blocking issue.

It will handle renamed or moved files if they're added to git's index.

The trade-off here is whether to rely on the file system's state or
the scm's state when making the decision of whether we should cache a
build file. I decided to go with the scm because there's an action
associated with changed and unchanged source units: changed source
units need to be rebuilt, and unchanged source units need to be
copied. Relying on a makefile makes sense when the choice is between
action and inaction. Additionally, we want srclib to know what
source units have changed when we start rebuilding unchanged source
units that depend on units that have changed, and to check if a cached
build file exists before copying it. I think that relying on the scm
for this information is more clear.

If we use globs to see if a makefile has changed, then we're relying
on every toolchain to pass us the correct globs (which isn't a big
deal, but I think that doing less work with each individual toolchain
is better reliability's sake), and we're using the state of the
filesystem to tell us whether a file changed, instead of the scm. I
don't think that's a big deal for builds on our site, because the
behavior will be the same using either method, but for using srclib
interactively, "srclib builds every unit that changed since the last
built commit" is more clear to me than "srclib builds every unit since
the last time it ran". If a build of source unit can fail based on the
state of another source unit, then it's less reliable to rebuild based
on the file's modification time.

I think getting the state of a repo from the scm makes more sense, but
I'm open to the idea of using globs and mtimes instead if there's a
convincing case for it.

This branch will do the right thing when given your example if the new
file is added to git's index (once I finish responding to your comments).